### ISSUE TITLE
[BUG][Spark] `maxBytesPerTrigger` and `maxFilesPerTrigger` not working

### ISSUE THREAD

**Issue by jasperp-be** [delta-io/delta#2268]
#### Which Delta project/connector is this regarding?
- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel

### Describe the problem
Adding the option maxFilesPerTrigger or maxBytesPerTrigger doesn't seem to do anything.

I have a streaming reading from and writing to delta. At the beginning, when `numBytesOutstanding` is very low, the performance is really good. The batch duration and processing rate/second are what I expect. Once the `numBytesOutstanding` increased (Due to upscaling upstream stream), the performance drops a lot (batch_duration and proc rate). I assume this is because I was not limiting the batch size enough. However, when I do this with `maxBytesPerTrigger` and `maxFilesPerTrigger`, I don't see a difference.

I used extreme values with the following:
```
df = (
    spark.readStream
    .option("maxFilesPerTrigger", 4)
    .option("maxBytesPerTrigger", "1m")
    .format("delta")
    .load(input_folder)
)
```
I expected this would result in a much lower batch_duration, yet, it still stays stuck. While it is still working on the first batch, in the SQL/DataFrame tab of the Spark UI, I find the following metrics of the Scan parquet step
![image](https://github.com/delta-io/delta/assets/34094863/c91e1c41-eea4-4495-89a7-79e1c9930e94)
The `number of files read` is exactly 1000 (The default value of `maxFilesPerTrigger`), which makes me assume it doesn't do anything.

In the stage of a batch, the Input Size is also very large (A lot larger than the 1MB I would expect)
<img width="146" alt="image" src="https://github.com/delta-io/delta/assets/34094863/5b722884-8624-4775-a62f-3f4c38686a75">
This screenshot is with metrics when only finished 32/155 stages.

My transformations are quite complex (Mainly a pandas UDF doing API requests), but I wouldn't expect to impact this.
<img width="102" alt="image" src="https://github.com/delta-io/delta/assets/34094863/f9405952-b60a-4e14-98c8-bcb9b69e90b7">

#### Steps to reproduce
See above

#### Observed results
See above

#### Expected results
Smaller size/less files processed per batch

#### Further details
Running this on Databricks DBR 14.1
5x workers of 32GB memory, 4 cores. Driver 26GB, 4 cores. No Photon

### Environment information

* Delta Lake version: 3.0.0
* Spark version: 3.5.0
* Scala version: 2.12.15
* [Databricks Runtime System Config](https://docs.databricks.com/en/release-notes/runtime/14.1.html#system-environment)

### Willingness to contribute

The Delta Lake Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the Delta Lake code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [x] Yes. I would be willing to contribute a fix for this bug with guidance from the Delta Lake community.
- [x] No. I cannot contribute a bug fix at this time.

Willing to contribute, but not sure if enough knowledge.

**Comment by tdas** [comment]
"I assume this is because I was not limiting the batch size enough. However, when I do this with maxBytesPerTrigger and maxFilesPerTrigger, I don't see a difference."

If there is a growing backlog of data yet to be processed (that is, high `numBytesOutstanding`), then larger batches with more data in each of them is likely to increase the throughput and processing efficiency. There is some amount of per-batch fixed overheads and so lower number of larger batches reduces the overall overheads. 

So while it is worth looking into why maxFilePerTrigger setting is not kicking in ... it may not be the best policy to use it for your use case.

**Comment by jasperp-be** [comment]
I agree with your logic. However, in my case, the batch duration suddenly increases a lot (from <1min to +20min to currently not being able to process the batch in a reasonable time). The processing rate also decreases in this case (which I find also counterintuitive, but is maybe because my specific transformation logic)

**Comment by jasperp97** [comment]
Update from my side: The `maxBytesPerTrigger` and `maxFilesPerTrigger` seem to work, but not for the first batch of the stream. Is this expected behaviour or not? If so, is this somewhere in the documentation, I didn't find it anywhere?

The issue is that the `maxFilesPerTrigger` default of 1000 is way too large. I kept getting OOM errors. I now decreased the `spark.sql.files.maxPartitionBytes` to +- 40MB, which made it pass the initial batch.

**Comment by jrampel** [comment]
I also encountered this issue, which is quite problematic.

My code is a simple scala job using `readStream`, `writeStream`, a `forEachBatch` and `AvailableNow` trigger.
My forEachBatch function performs a lot of updates in a delta table, which makes it very sensible to the size of the batches. 
A batch of 1000 files (the default value) is too much to handle and fails.

Despite trying many things, I could never get delta to apply the `maxBytesPerTrigger` or `maxFilesPerTrigger` options.

I do have a (dirty) workaround though.

I overwrite the default value for `maxFilesPerTrigger` in `DeltaOptions` using java/scala reflection.
https://github.com/delta-io/delta/blob/b721e1913f5bdf93e69254333d839fa32673a30a/spark/src/main/scala/org/apache/spark/sql/delta/DeltaOptions.scala#L249

With this overwritten, you need to keep `maxBytesPerTrigger` empty because of this condition in `DeltaSource`
https://github.com/delta-io/delta/blob/b721e1913f5bdf93e69254333d839fa32673a30a/spark/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSource.scala#L1260-L1265

This does work for now, but if anyone has a cleaner workaround or a proper patch, that'd be great.
`forEachBatch` is very convenient because of checkpoints.

I did some debugging but I don't see any obvious fix, it really seems like the option value is not routed to `DeltaSource.AdmissionLimits`. I'll look into it further if I have the time.

**Comment by foss-contributor** [comment]
@jasperp97, thank you for your contribution! This issue has been inactive for over 18 months and will be closed in one week. To keep it open, please leave a comment on this issue or contact us in the `#contributors` channel on [Delta Lake Slack](https://go.delta.io/slack). We appreciate your involvement in Delta Lake!
